activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1514,6 +1572,35 @@ def build_select(arel)
         end
       end
 
+      def build_with(arel)
+        return if with_values.empty?
+
+        with_statements = with_values.map do |with_value|
+          case with_value
+          when Arel::Nodes::As, Arel::Nodes::TableAlias then with_value
+          when Hash then build_with_value_from_hash(with_value)
+          else
+            raise ArgumentError, "Unsupported argument type: #{with_value} #{with_value.class}"
+          end
+        end
+
+        arel.with(with_statements)
+      end
+
+      def build_with_value_from_hash(hash)
+        hash.map do |name, value|
+          expression =
+            case value
+            when Arel::Nodes::SqlLiteral then Arel.sql("(#{value})")
```

Comment:
Use a Grouping node to add parens, don't re-create the SqlLiteral by modifying its string form.